src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -598,6 +602,32 @@ public Task WriteAsync(ArraySegment<byte> data, CancellationToken cancellationTo
             }
         }
 
+        private void VerifyAndUpdateWrite(int count)
+        {
+            TryInitializeResponseContentLength();
+
+            if (_responseContentLength.HasValue && _responseBytesWritten + count > _responseContentLength.Value)
+            {
+                _keepAlive = false;
+                throw new InvalidOperationException(
+                    $"Response Content-Length mismatch: too many bytes written ({_responseBytesWritten + count} of {_responseContentLength.Value}).");
+            }
+
+            _responseBytesWritten += count;
+        }
+
+        protected void TryInitializeResponseContentLength()
+        {
+            if (!_responseContentLength.HasValue)
+            {
+                var responseHeaders = FrameResponseHeaders;
+                if (responseHeaders != null && responseHeaders.HasContentLength)
+                {
+                    _responseContentLength = long.Parse(responseHeaders.HeaderContentLength.ToString(), NumberStyles.None);
```

Comment:
We check for its presence but we don't try to parse it.

I disagree with you on this point. We should throw if the `Content-Length` header is set to an invalid value. The whole point of this change is to track errors related to response `Content-Length`, this would be one of them. Without a valid value, we can't track it.